ujt/transformers.py

```diff
@@ -28,38 +35,142 @@ def apply_highlighted_edge_class_to_elements(elements, user_journey_name):
-        edges_map[(client_name,
-                   dependency.target_name
-                  )]["classes"] = constants.HIGHLIGHTED_UJ_EDGE_CLASS
-
-    while node_frontier:
-        node = node_frontier.pop()
-        for dependency in node.dependencies:
-            node_frontier.append(node_name_message_map[dependency.target_name])
-            edges_map[(node.name,
-                       dependency.target_name
-                      )]["classes"] = constants.HIGHLIGHTED_UJ_EDGE_CLASS
+    """ Applies the highlighted edge class to edges within a specific user journey.
+
+    Traverses the edges map (gneerated from the elements array) instead of the underlying
+    protobufs, in order to easily support virtual nodes.
+
+    Args:
+        edges_map: A dictionary mapping edge sources to a list of cytoscape elements describing edges originating from the source
+        user_journey_name: The fully qualified user journey name to highlight.
+
+    Returns:
+        An updated dictionary mapping edge sources to a list of edge cytoscape elements, with the new highlighted class applied.
+    """
+    node_frontier_names = deque([user_journey_name])
+    while node_frontier_names:
+        source_name = node_frontier_names.popleft()
+        for edge in edges_map[source_name]:
+            edge["classes"] = constants.HIGHLIGHTED_UJ_EDGE_CLASS
+            node_frontier_names.append(edge["data"]["target"])
 
     return edges_map
+
+
+def apply_virtual_nodes_to_elements(elements):
+    virtual_node_map = state.get_virtual_node_map()
+    parent_virtual_node_map = state.get_parent_virtual_node_map()
+
+    new_elements = []
+    for element in elements:
+        if utils.is_node_element(element):
+            highest_collapsed_virtual_node_name = utils.get_highest_collapsed_virtual_node_name(
+                element["data"]["ujt_id"])
+            if highest_collapsed_virtual_node_name is None:
+                # not within collapsed node
+                if element["data"][
+                        "id"] in parent_virtual_node_map and "parent" not in element[
+                            "data"]:
+                    element["data"]["parent"] = parent_virtual_node_map[
+                        element["data"]["ujt_id"]]
+                new_elements.append(element)
+
+        else:
+            new_source = utils.get_highest_collapsed_virtual_node_name(
+                element["data"]["source"])
+            new_target = utils.get_highest_collapsed_virtual_node_name(
+                element["data"]["target"])
+
+            if new_source is not None:
+                element["data"]["source"] = new_source
+            if new_target is not None:
+                element["data"]["target"] = new_target
+
+            element["data"][
+                "id"] = f"{element['data']['source']}/{element['data']['target']}"
+
+            if element["data"]["source"] != element["data"]["target"]:
+                new_elements.append(element)
```

Comment:
Here we transform cytoscape elements to account for the currently collapsed virtual nodes. We remove the nodes hidden within a virtual node. Moreover, we adjust the affected edges that originally connected to nodes within the virtual node, to point to the virtual node itself.